File: backend/services/settings_service.py
import json
import os

SETTINGS_FILE = "storage/settings.json"

DEFAULT_SETTINGS = {
    "show_qrcode": True,
    "remote_enabled": True,
    "selected_logo": "default",
    "strip_background": "#1a1a2e",
    "strip_background_image": None,
    "strip_layout": {
        "cols": 2,
        "rows": 2,
        "cells": [
            {
                "id": 1,
                "type": "logo",
                "logo": "default",
                "col": 0,
                "row": 0,
                "col_span": 1,
                "row_span": 1
            },
            {
                "id": 2,
                "type": "photo",
                "photo_index": 0,
                "col": 1,
                "row": 0,
                "col_span": 1,
                "row_span": 1
            },
            {
                "id": 3,
                "type": "photo",
                "photo_index": 1,
                "col": 0,
                "row": 1,
                "col_span": 1,
                "row_span": 1
            },
            {
                "id": 4,
                "type": "photo",
                "photo_index": 2,
                "col": 1,
                "row": 1,
                "col_span": 1,
                "row_span": 1
            }
        ]
    }
}


class SettingsService:

    def get(self) -> dict:
        if not os.path.exists(SETTINGS_FILE):
            self.save(DEFAULT_SETTINGS)
            return DEFAULT_SETTINGS.copy()
        with open(SETTINGS_FILE, "r") as f:
            return json.load(f)

    def save(self, data: dict) -> dict:
        os.makedirs(os.path.dirname(SETTINGS_FILE), exist_ok=True)
        with open(SETTINGS_FILE, "w") as f:
            json.dump(data, f, indent=2)
        return data

    def reset(self) -> dict:
        return self.save(DEFAULT_SETTINGS.copy())

    def update(self, partial: dict) -> dict:
        current = self.get()
        current.update(partial)
        return self.save(current)

    def get_photos_count(self) -> int:
        """Retourne le nombre de cellules photo dans le layout."""
        settings = self.get()
        cells = settings.get("strip_layout", {}).get("cells", [])
        return len([c for c in cells if c.get("type") == "photo"])

File: frontend/src/app/settings/settings.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { ConfigService } from '../shared/config.service';

export interface Cell {
  id: number;
  type: 'logo' | 'photo' | 'empty';
  logo?: string;
  photo_index?: number;
  col: number;
  row: number;
  col_span: number;
  row_span: number;
}

export interface StripLayout {
  cols: number;
  rows: number;
  cells: Cell[];
}

export interface AppSettings {
  show_qrcode: boolean;
  remote_enabled: boolean;
  selected_logo: string;
  strip_background: string;
  strip_background_image: string | null;
  strip_layout: StripLayout;
  storage_quota_mb: number;
}

@Injectable({ providedIn: 'root' })
export class SettingsService {
  isExporting = false;

  constructor(private http: HttpClient, private config: ConfigService) { }

  private get apiUrl(): string {
    return `${this.config.apiUrl}/api/settings`;
  }

  private get storageApiUrl(): string {
    return this.config.apiUrl;
  }

  getSettings(): Observable<AppSettings> {
    return this.http.get<AppSettings>(this.apiUrl);
  }

  saveSettings(data: AppSettings): Observable<AppSettings> {
    return this.http.post<AppSettings>(this.apiUrl, data);
  }

  resetSettings(): Observable<AppSettings> {
    return this.http.post<AppSettings>(`${this.apiUrl}/reset`, {});
  }

  getLogos(): Observable<{ logos: string[] }> {
    return this.http.get<{ logos: string[] }>(`${this.apiUrl}/logos`);
  }

  uploadLogo(file: File): Observable<any> {
    const fd = new FormData();
    fd.append('file', file);
    return this.http.post(`${this.apiUrl}/logo`, fd);
  }

  deleteLogo(filename: string): Observable<any> {
    return this.http.delete(`${this.apiUrl}/logo/${encodeURIComponent(filename)}`);
  }

  getBackgrounds(): Observable<{ backgrounds: { filename: string; url: string }[] }> {
    return this.http.get<any>(`${this.apiUrl}/backgrounds`);
  }

  uploadBackground(file: File): Observable<any> {
    const fd = new FormData();
    fd.append('file', file);
    return this.http.post(`${this.apiUrl}/background`, fd);
  }

  deleteBackground(filename: string): Observable<any> {
    return this.http.delete(`${this.apiUrl}/background/${encodeURIComponent(filename)}`);
  }

  async exportPhotos(token: string): Promise<void> {
    const url = `${this.apiUrl}/export`;
    const res = await fetch(url, {
      headers: { Authorization: `Bearer ${token}` }
    });
    const blob = await res.blob();
    const a = document.createElement('a');
    a.style.display = 'none';
    a.href = URL.createObjectURL(blob);
    a.download = 'photobooth_photos.zip';
    document.body.appendChild(a);
    a.click();
    setTimeout(() => {
      document.body.removeChild(a);
      URL.revokeObjectURL(a.href);
    }, 100);
  }

  shutdown(): Observable<any> {
    return this.http.post(`${this.apiUrl}/shutdown`, {});
  }

  getLogoUrl(logo: string): string {
    if (logo === 'default') return '';
    return `${this.storageApiUrl}/logos/${logo}`;
  }

  getBackgroundUrl(filename: string): string {
    return `${this.storageApiUrl}/backgrounds/${encodeURIComponent(filename)}`;
  }
}